spring-cloud-gcp-pubsub/src/main/java/com/google/cloud/spring/pubsub/core/PubSubEventSpecificProperties.java

```diff
@@ -0,0 +1,490 @@
+/*
+ * Copyright 2021-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spring.pubsub.core;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.google.api.gax.batching.FlowController.LimitExceededBehavior;
+
+/**
+ * Properties for Publisher or Subscriber specific configurations.
+ *
+ * @author João André Martins
+ * @author Mike Eltsufin
+ * @author Chengyuan Zhao
+ */
+public class PubSubEventSpecificProperties {
+
+	private final Map<String, Properties> properties = new HashMap<>();
+
+	/**
+	 * Contains settings specific to the subscriber factory.
+	 */
+	private final Subscriber subscriber = new Subscriber();
+
+	/**
+	 * Contains settings specific to the publisher factory.
+	 */
+	private final Publisher publisher = new Publisher();
+
+	public Subscriber getSubscriber() {
+		return this.subscriber;
+	}
+
+	public Publisher getPublisher() {
+		return this.publisher;
+	}
+
+	public Map<String, Properties> getProperties() {
+		return this.properties;
+	}
+
+	public Subscriber getSubscriber(String name) {
+		if (!this.properties.containsKey(name)) {
+			Properties property = new Properties();
+			property.setSubscriber(this.subscriber);
+			this.properties.put(name, property);
+		}
+
+		return this.properties.get(name).getSubscriber();
+	}
+
+	public static class Properties {
```

Comment:
I think calling this class "Properties" could be too ambiguous. "PubSubProperties" or "PubSubConfiguration"? Also, I think it could be a standalone class.

Reply:
Thanks for the suggestions! I agree that the naming is a little vague, I will clean this up. 

> Do we actually need the intermediate class `Properties`? It is just a container for a single `Subscriber` property bag. 

That is a great question! Having the ui be `spring.cloud.gcp.pubsub.*name*.subscriber.executor-threads` instead of having the intermediate `Properties` class would certainly be ideal. However, this didn't play out very well with Spring autoinjection. When we had the setup in the following manner the global configurations were working fine but for the subscription-specific properties, we were relying on the default values instead of the ones being set. The map properties weren't getting recognized.
```
        private final Map<String, Subscriber> properties = new HashMap<>();

	/**
	 * Contains settings specific to the subscriber factory.
	 */
	private final Subscriber subscriber = new Subscriber();
``` 
Looking at this [spring reference doc](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config.typesafe-configuration-properties.merging-complex-types), we may still need an intermediate class?

Reply:
Yep, you are right. Let's keep the `.properties.`